Vectorized Dynamics for Fast Parallel Simulation on CPU or GPU

Hersh500 opened the pull request:
# Summary
This PR implements vectorized (batched) versions of the Multirotor dynamics, SE(3) Controller, and MinSnap trajectories in PyTorch. This allows fast simulation of multiple vehicles in parallel on the CPU or GPU. With larger batch sizes, the obtained speed up on CPU (single-threaded) can range from 10-16x compared to simply simulating all drones sequentially (~400-600 FPS), and on GPU the speedup can range from 10x-**66x**. On my machine, on GPU, I was able to simulate 20000 drones in parallel for an amortized 40k FPS across the simulation episode. With an optional simplified integration scheme (rk4), I achieved a **100x** speedup on CPU (60k FPS) and **150x** speedup on GPU (90k FPS).
![CPU_vs_GPU_rotorpy_batch_perf_dopri5](https://github.com/user-attachments/assets/7201e1a9-bce6-4f26-a17b-3ff876787641)
![CPU_vs_GPU_rotorpy_batch_perf_rk4](https://github.com/user-attachments/assets/8137ca4b-4a53-423b-8bb9-075a76e639a2)


# New Features
## Classes
- `BatchedMultirotor`: a class that implements vectorized MultiRotor dynamics, with a similar interface to the standard `Multirotor`. If desired, it also allows the user specify which drones in the batch to step (which is useful if certain drones in the batch terminate before others). It also allows the user to specify using an integrator with an adaptive step size or fixed step size. The ODE solver, implemented fully in PyTorch is an open source package: https://github.com/rtqichen/torchdiffeq . **Currently, all drones in the batch must have the same quad_params.**
- `minsnap.BatchedMinSnap`: a class that computes flat outputs from multiple MinSnap trajectories at the same time using vectorized ops. It supports specifying a different timestep for each trajectory in the batch.
- `quadrotor_control.BatchedSE3Control`: a class that implements a vectorized SE3 controller, with a similar interface to `quadrotor_control.SE3Control`. Similarly to BatchedMultirotor, it allows specifying which drones in the batch to compute control inputs for. It also allows the user to specify different feedback gains for each drone in the batch, if desired. _Fails a test right now because its API is different than `SE3Control`_

## Utility Functions
- `batch_simulate.py` contains `simulate_batch()`, an analogue to the standard `simulate()` function, with a slightly different API. Using the vectorized classes above, it simulates multiple drones in parallel. Each drone can have a different reference trajectory and can terminate at a different time than the others, in which case the "state", "flat", and "control" outputs for that drone will be 0 for the remaining timesteps.
- `utils/trajgen_utils.py`: various utilities for generating random `MinSnap` trajectories.

## Examples
- `examples/batched_simulation.py`: a walkthrough of how to use the new classes, and compares the outputs of the batched simulation to the sequential sim.
- `examples/benchmark_batched_simulation.py`: measures the speed (FPS and wall-clock time) and memory usage of the sequential simulation, CPU-batched simulation, and GPU-batched simulation. 

## Tests
- `tests/test_batched_sims.py`: Checks that the outputs of the new batched classes are equal to the existing rotorpy classes, if you called them individually for each drone in the batch.

# Changes to Existing Code
- `simulate.py`: added a `print_fps` option (default False) to print the FPS achieved by the simulation at each timestep.
- `wind/default_winds.py`: added an option to specify `num_drones` to `NoWind` and `ConstantWind` -- if `num_drones` is greater than one, they return torch tensors with a batch dimension.

# Fidelity
### dopri5
By visual inspection (and by the test code specified above), the batched dynamics with the `dopri5` solver match the existing dynamics (with `motor_noise=0`), even for aggressive trajectories:
X,Y,Z positions:
![dopri5_pos](https://github.com/user-attachments/assets/353dde4d-e987-4f36-ae12-9134b776f7ec)
RPY Body Rates:
![dopri5_bodyrates](https://github.com/user-attachments/assets/2d9072c8-3f90-44d2-b001-718ae7dc3bd8)

### rk4
With the `rk4` solver, the simulation remains stable, but differ from the existing implementation:
![rk4_pos](https://github.com/user-attachments/assets/77a9ce64-be76-40e9-8e3b-88590287b7bd)
![rk4_bodyrates](https://github.com/user-attachments/assets/fd330a3f-4751-46e5-8841-ed3b44c126af)

spencerfolk commented:
This is so cool. The speedup is more than I could have ever imagined!! Thanks for the thorough PR with included benchmarks. 

After our discussion we came up with the following to-dos: 

To-Do for you:
- [ ] Test with a `numpy 1.x` version. Tests for Python 3.8 and 3.9 are failing because I don't think `numpy 2.x` supports these versions. If for some reason `numpy 1.x` is causing a big slowdown, we'll talk about potentially removing support for older Python versions or at least not supporting batched environments for these versions. 
- [ ] Implement more control abstractions besides `cmd_motor_speeds` in `BatchedMultirotor` for parity with the existing `Multirotor` class.
- [ ] Add support for different `quad_params` for each drone in the batch. 

To-Do for me:
- [ ] Remove `motor_noise_std` parameter, or at least set it to 0 by default to avoid annoying discrepancies in potential future benchmarking or tests. 
- [ ] Implement batched versions of the other default winds. 
- [ ] Fix the controller input signature tests (currently failing for your PR) to be more permissive. 
- [ ] Start updating necessary documentation. 
- [ ] Begin thinking about necessary modifications to the existing `gymnasium` environment to use the `BatchedMultirotor` vehicle. 
- [ ] Run your benchmarking scripts on my machines to compare performance gains across different systems.

Hersh500 commented:
Made a bunch of changes!

1. changed `pyproject.toml` to `numpy>=1.25.0` and `torch>=1.11.0`, which should work with `python3.8`. I only did cursory testing, but it doesn't seem like it obviously affects the performance. Also split `opt-einsum` into its own dependency since it was not bundled with older versions of `torch`.
2. Drones that terminate early will now have `NaN` in the `state` arrays after they terminate. This shouldn't affect simulator functionality, but hopefully will make it more obvious if you try to index into a drone after it has already terminated.
3. You can now specify different quadrotor parameters for each drone in the batch. The machinery for doing this makes the code a little more clunky (as you now have to have a list of `quad_params` dictionaries instead of just one), but is obviously very powerful as you can simulate many different drones simultaneously (e.g. for domain randomization). **For performance, this seems to decrease the simulation speed at larger batch sizes. Running the benchmark script -- on CPU with `dopri5`, the FPS at 20k batch size goes from ~9000 to ~7500, and on GPU, FPS at 20k batch size goes from ~39000 to ~35000. However, I think the increase in flexibility might be worth the hit to simulation speed.**
5. I created a class called `BatchedDynamicsParams` in `batched_multirotor.py` which acts as a container class for the batch of drone model parameters, that can be shared across the `BatchedSE3Control` and `BatchedMultirotor` classes.
6. Added the higher level control abstractions to `BatchedMultirotor`, and tested their functionality in `tests/test_batched_sim.py` . This makes `tests/test_batched_sim.py` kind of computationally heavy though -- I'm not sure if there's a limit in terms of github actions.

spencerfolk commented:
@Hersh500  This PR is definitely ready to merge, but I'm thinking we should definitely prep this for a `v2.0.0` release. Here are the things we (but mostly me) need to do for this. 

- [ ] Implement a `BatchedEnvironment` class that mimics the existing [`Environment`](https://github.com/spencerfolk/rotorpy/blob/main/rotorpy/environments.py) but uses your `batch_simulate()` function. 
- [ ] Incorporate `BatchedMultirotor` into a new `gymnasium` environment. Put together new PPO hover training demonstration. 
- [ ] Add batched versions of all given wind profiles. 
- [ ] Update documentation where necessary across all modules and especially the main README to highlight batched sims. 
- [ ] Clean up benchmarking scripts and produce clean summary plots that can be highlighted in the main README.

spencerfolk reviewed (approved):
Super impressive! Thanks for incorporating these changes!
